src/client/pool/singleton.rs

```diff
@@ -0,0 +1,333 @@
+//! Singleton pools
+//!
+//! The singleton pool combines a MakeService that should only produce a single
+//! active connection. It can bundle all concurrent calls to it, so that only
+//! one connection is made. All calls to the singleton will return a clone of
+//! the inner service once established.
+//!
+//! This fits the HTTP/2 case well.
+
+use std::sync::{Arc, Mutex};
+use std::task::{self, Poll};
+
+use tokio::sync::oneshot;
+use tower_service::Service;
+
+use self::internal::{DitchGuard, SingletonError, SingletonFuture};
+
+type BoxError = Box<dyn std::error::Error + Send + Sync>;
+
+/// A singleton pool over an inner service.
+///
+/// The singleton wraps an inner service maker, bundling all calls to ensure
+/// only one service is created. Once made, it returns clones of the made
+/// service.
+#[derive(Debug)]
+pub struct Singleton<M, Dst>
+where
+    M: Service<Dst>,
+{
+    mk_svc: M,
+    state: Arc<Mutex<State<M::Response>>>,
+}
+
+#[derive(Debug)]
+enum State<S> {
+    Empty,
+    Making(Vec<oneshot::Sender<S>>),
```

Comment:
this should be bounded in some manner! if the `SingletonFuture` gets stuck whilst establishing a connection, we could end up with an unbounded number of waiters here.

it'd be nice if configuring that was optional (_e.g. with a builder-style method?_) with a relatively high default value.

ideally, we also introduce some lightweight test coverage to ensure that only N waiters can enter the queue once we've begun "making" the inner service.

Reply:
I thought about if it needed bounding, but figured bounding it also something that can happen with an additional layer, just slotting in `ConcurrenyLimit` layer, for instance. In that case, it feels like the appropriate thing for this layer is to act just like a vec, where the bound is just memory.

Reply:
that makes sense to me. thank you, i will mark this thread as resolved!